tests/test_md.py

```diff
@@ -640,6 +640,127 @@ def test_multiple_steps_local_no_free_particles(self):
+
+        temperature = 300
+        dt = 1.5e-3
+        friction = 0.0
+        radius = 1.2
+
+        # Select a single particle to use as the reference, will be frozen
+        local_idxs = np.array([len(coords) - 1], dtype=np.int32)
+
+        v0 = np.zeros_like(coords)
+        bps = [nb_pot.bind(params).bound_impl]
+
+        reference_values = []
+        for bp in bps:
+            reference_values.append(bp.execute(coords, box))
+
+        intg = LangevinIntegrator(temperature, dt, friction, masses, seed)
+
+        steps = 10
+
+        # Verify that initializing local md doesn't modify global md behavior
+        ctxt = custom_ops.Context(coords, v0, box, intg.impl(), bps)
+        ref_xs, ref_boxes = ctxt.multiple_steps(steps)
+
+        ctxt = custom_ops.Context(coords, v0, box, intg.impl(), bps)
+        ctxt.initialize_local_md()
+        comp_xs, comp_boxes = ctxt.multiple_steps(steps)
+
+        np.testing.assert_array_equal(ref_xs, comp_xs)
+        np.testing.assert_array_equal(ref_boxes, comp_boxes)
+
+        # Verify that initializing local md doesn't modify global md behavior
+
+        ctxt = custom_ops.Context(coords, v0, box, intg.impl(), bps)
+        ref_local_xs, ref_local_boxes = ctxt.multiple_steps(steps)
+
+        ctxt = custom_ops.Context(coords, v0, box, intg.impl(), bps)
+        ctxt.initialize_local_md()
+        comp_local_xs, comp_local_boxes = ctxt.multiple_steps(steps)
+
+        np.testing.assert_array_equal(ref_local_xs, comp_local_xs)
+        np.testing.assert_array_equal(ref_local_boxes, comp_local_boxes)
+
+    def test_local_md_with_selection_mask(self):
+        """Verify that running local md with a selection mask works as expected"""
+        seed = 2023
+        rng = np.random.default_rng(seed)
+
+        N = 8
+        D = 3
+
+        coords = rng.random(size=(N, D)).astype(dtype=np.float64) * 2
+        box = np.eye(3) * 3.0
+        masses = rng.random(N)
+
+        E = 2
+
+        params, potential = prepare_nb_system(
+            coords,
+            E,
+            p_scale=3.0,
+            cutoff=1.0,
+        )
+        nb_pot = potential.to_gpu(np.float32)
+
+        temperature = 300
+        dt = 1.5e-3
+        friction = 0.0
+        radius = 1.2
+
+        # Select a single particle to use as the reference, will be frozen
+        local_idxs = np.array([len(coords) - 1], dtype=np.int32)
+
+        v0 = np.zeros_like(coords)
+        bps = [nb_pot.bind(params).bound_impl]
+
+        reference_values = []
+        for bp in bps:
+            reference_values.append(bp.execute(coords, box))
+
+        intg = LangevinIntegrator(temperature, dt, friction, masses, seed)
+
+        idxs = np.arange(0, len(coords))
+
+        reference_idx = rng.choice(idxs)
+        # TODO, make it so you can't have the reference in the free
+        free_particles = rng.choice(idxs, size=len(coords) // 2)
+        free_particles = np.delete(free_particles, free_particles == reference_idx)
```

Comment:
👍 -- should assert 

Aside: I think `reference_idx` in `free_particles` would technically be permissible if `r0 == 0`.

Problem being avoided: A discontinuity in `log(1 - exp(-u_flat_bottom_restraint(r; r0; k))` if `reference_idx` moves in such a way that its distance `r` to any frozen particle that was initially `r > r0` away from `x[reference_idx]` becomes `r < r0` away. We prevent local updates from crossing this discontinuity by preventing local updates from moving `reference_idx`.

(Bonus aside: We also have an assertion in the wrappers that `r0 > 0.1` whose motivation is less clear to me -- if we relax this assertion, then we could technically allow `reference_idx` in `free_particles` as long as `r0 == 0.0`. But I see no practical motivation to allow this.)

Reply:
Asserted in  https://github.com/proteneer/timemachine/pull/1005/commits/37cd439f94c34fd532d9e286434259287b201144

Comment removed in https://github.com/proteneer/timemachine/pull/1005/commits/801bb6c51982b1c566726fd1d1062e2506d7a72f